lib/Onyx.d.ts

```diff
@@ -0,0 +1,278 @@
+import {Component} from 'react';
+import {PartialDeep} from 'type-fest';
+import * as Logger from './Logger';
+import {DeepRecord, Key, CollectionKey, Selector, Value} from './types';
+
```

Comment:
<I want to make a thread out of this discussion cc @hayata-suenaga @fabioh8010 >

> > > At the moment we don't have means to differentiate between ONYXKEYS.COLLECTION.REPORT and ${ONYXKEYS.COLLECTION.REPORT}${reportId} which would only return Report | null, causing mismatches between the Onyx value and the Onyx prop. I'm looking for a fix for that.
> > 
> > 
> > so currently, the onyx prop for `${ONYXKEYS.COLLECTION.REPORT}${reportId}` is inferred as `Record<string, Report | null>`?
> 
> @hayata-suenaga Let's consider this types:
> 
> ```ts
> const ONYXKEYS = {
>     ACCOUNT: 'account',
>     COLLECTION: {
>         REPORT: 'report_',
>     },
> } as const;
> 
> type DeepValueOf<T> = T extends object ? DeepValueOf<T[keyof T]> : T;
> 
> type OnyxKeysMap = typeof ONYXKEYS;
> type OnyxCollectionKey = `${ValueOf<OnyxKeysMap['COLLECTION']>}${string}`;
> type OnyxKey = DeepValueOf<Omit<OnyxKeysMap, 'COLLECTION'>>;
> 
> type Account = {
>     id: string;
>     name?: string;
> };
> 
> type Report = {
>     id: string;
>     isArchived: boolean;
>     data: {
>         message: string;
>         isRead?: boolean;
>     };
> };
> 
> type OnyxValues = {
>     [ONYXKEYS.ACCOUNT]: Account;
>     [report: `${typeof ONYXKEYS.COLLECTION.REPORT}${string}`]: Report;
> };
> ```
> 
> And this component:
> 
> ```tsx
> type OnyxProps = {
>     oneReport: Report | null;
>     allReports: Record<string, Report | null> | null;
> };
> 
> type Props = OnyxProps & {
>     requiredProp: string;
> };
> 
> function Component({requiredProp, oneReport, allReports}: Props) {
>     return null;
> }
> 
> const WithOnyxComponent = withOnyx<Props, OnyxProps>({
>     oneReport: {
>         /**
>          * CORRECT
>          * 
>          * "oneReport" is expected by the HOC typings to be `Report | null`,
>          * because I'm selecting only one report.
>          */
>         key: `${ONYXKEYS.COLLECTION.REPORT}${'id1'}`,
>     },
>     allReports: { // TS error
>         /**
>          * INCORRECT
>          * 
>          * "allReports" is still expected by the HOC typings to be `Report | null`,
>          * but it should expect `Record<string, Report | null> | null` because Onyx
>          * will give me an object with all the reports when using a collection key
>          * without suffix, like `ONYXKEYS.COLLECTION.REPORT`.
>          * 
>          * Because of that, I get an error because `allReports` prop expects
>          * `Record<string, Report | null> | null` but the HOC typings is expecting
>          * `Report | null` because it was declared in this way in `OnyxValues` type.
>          */
>         key: ONYXKEYS.COLLECTION.REPORT,
>     },
> })(Component);
> 
> function Playground() {
>     return <WithOnyxComponent requiredProp="some string" />;
> }
> ```
> 
> This is the issue. We need to figure out a way to make the HOC typings expect `Record<string, T> | null` from the mapping when using a collection key without suffix.

Reply:
I explored some options and figured a way to distinguish between the whole collection (`ONYXKEYS.COLLECTION.REPORT`) and collection item (`${ONYXKEYS.COLLECTION.REPORT}${'id1'}`). I'm not sure if it's possible to incorporate it into existing types... 

@fabioh8010 wdyt?

```ts
type Key = 'account';
type CollectionKey = 'download_' | 'report_'
type OnyxKey = Key | CollectionKey

type Values = {
    account: boolean
    [download: `download_${string}`]: number;
    [report: `report_${string}`]: string;
}

type GetOnyxValue<K extends Key | `${CollectionKey}${string}`> = K extends `${OnyxKey}${string}`
        ? K extends `${OnyxKey}`
            ? Record<string, Values[K]>
            : Values[K]
        : never;

type Account = GetOnyxValue<'account'>
type DownloadAll = GetOnyxValue<'download_'>
type DownloadOne = GetOnyxValue<'download_whatever'>
type ReportAll = GetOnyxValue<'report_'>
type ReportOne = GetOnyxValue<'report_1'>

type Error1 = GetOnyxValue<'account_test'>
type Error2 = GetOnyxValue<'asdasdsad'>
```

[TS playground](https://www.typescriptlang.org/play?ts=4.8.4#code/C4TwDgpgBA0hJQLxQOQEMDGGD2BXAdsCgNwBQokUAwtgDa0QbACW2+cCyKAJtgO75a2NNwD6KKAB9UAJwhhsM4OPLhoAeXwgAHhySx4U6nQZNW7eKVWUAamlq4IAZ30BvUlE9RMOAsABcUABG2HQQaPgeXgDavAJCIoEABnGCwmIAJK5OwDLM+ADmAL5JALqB+LgAtkEQMmReUNFyCkrJLYrKWTl5hSXlUD35BWRFVhTQAOIQwJo6dg4QADwwUBDawBD43C560klZNPSMLGwcRd25wyUAfPqr65vbLgeuc7rwF9lXfUlRjZ4APywNYbLY7KCvd7nP4AuFQYEAJUYim4SyGhQANFAFo4nNEYKUbv94Z5ArjnATSiTGhUIAA3OpkazQACCWDwhH001mWm0FKW6A5fhQxImUAAIvw0iJWfRuTN3gKeNKEmJRSzJar0ppoMgeUr7I5Bak1aI+AALNCbRkyDXi5GtYBy2gK3nzI3LFAdJTiMVqKCOzq6t2GxaCn3KACMGs1AFEZDJFFHQ3zlT5OcpNjl7QGE0mZAAmVMe8PoJzcNAVpwiUVAA)

Reply:
omg such a nice solution 🤩 

Couple of things I noticed:

- The `Account` type is derived incorrectly.
  <img width="350" alt="Screenshot 2023-07-20 at 10 25 02 AM" src="https://github.com/Expensify/react-native-onyx/assets/98560306/3fc2e8b3-bea0-4210-8295-3d42e9774bb9">

We we can make `GetOnyxValue` specific to collection keys.

```ts
type OnyxCollectionValue<K extends `${CollectionKey}${string}`> = K extends CollectionKey ? Record<string, Values[K]> : Values[K];
```

[TS playground](https://www.typescriptlang.org/play?ts=4.8.4#code/C4TwDgpgBA0hJQLxQOQEMDGGD2BXAdsCgNwBQokUAwtgDa0QbACW2+cCyKAJtgO75a2NNwD6KKAB9UAJwhhsM4OPLhoAeXwgAHhySx4U6nQZNW7eKVWUAamlq4IAZ30BvUlE9RMOAsABcUABG2HQQaPgeXgDavAJCIoEABnGCwmIAJK5OwDLM+ADmAL5JALqB+LgAtkEQMmReUNFyCkrJLYrKWTl5hSXlUD35BWRFVhTQAOIQwJo6dg4QADwwUBDawBD43C5JWTT0jCxsHEXducMlAHz6q+ub2y4HpscWCAD8UABKjIrcS0NCgAaKALRxOaIwUo3QJg5yQ0pkazQAAi-DSIgAgvR9NNZlptHCljx0QkxCgrsioGj4ulNNBkHi5oT7I5iakyaI+AALNCbABudQpVJ+rWA2NouJmzKJKA6SnElIm33knXpUvx81ZyzlqoVAEZhVSAKIyGSKAAMGpl2uJPjwhGFytN5pk+utBNl9r8ok2OSdaigLsUACYPVrFnanNw0NGnCIKUA)

Reply:
@blazejkustra @hayata-suenaga Thanks for the suggestions, checking 💪

Reply:
@fabioh8010 @hayata-suenaga I fixed it so it works for both `Key` and `CollectionKey`, [ts playground](https://www.typescriptlang.org/play?pretty=false&ts=4.8.4#code/C4TwDgpgBA0hJQLxQOQEMDGGD2BXAdsCgNwBQokUAwtgDa0QbACW2+cCyKAJtgO75a2NNwD6KKAB9UAJwhhsM4OLIVoAeXwgAHhySx4U6nQZNW7eGXLhoANTS1cEAM76A3qSheomHAWAAXFAARth0EGj4ZN5QANq8AkIiQQAGCYLCYgAkbs7AMsz4AOYAvikAukH4uAC2wRAy0d6xcgpKqa2Kyjl5BcVllVC9hUVkJVZqUADiEMCaOvaOEAA8MFAQ2sAQ+NyuetIpOTT0jCxsHCU9+SNlAHz6axtbO3uGAPxQi07OsTDlUEFHpttrsoIc3MdTGcLCAylAPgAlRiKbjLYbFAA0nwc31+5XuQS+LjxExsUAAglg8IR9DM5lptETlugqf4ULdVGSACL8DIicn0WmzeaMnErHi8pJidmcyg8xKZTTQZB0kVMiUKkSiPgACzQWwAbg0ZdZKEi2sABbQhfSFmLmZ0lOIOaboOaukqbWr7ShHcoAIwm11QACiMhkin9XoZ6t81OUWzyJsmYYjMgATNG7UtmWhnNw87sRCagA)

Reply:
@hayata-suenaga I was working together with @blazejkustra today to solve this problem, we are still stuck on this but we had some ideas today which I will try tomorrow.

Reply:
@hayata-suenaga Please see this https://github.com/Expensify/react-native-onyx/pull/267#issuecomment-1656430774